### PR TITLE
Reland [AMDGPU] Add AMDGPU specific variadic operation MCExprs

### PR DESCRIPTION
**Pull request by JanekvO**
Adds AMDGPU specific variadic MCExpr operations 'max' and 'or'. Required for moving function/program resource usage information propagation to MC layer as propagation of the maximum and logic OR values will go through MCExprs. Propagation of said resource usage info through these MCExpr operations compared to current handling in AMDGPUResourceUsageAnalysis will ease the constraint on the order for a function's resource usage analysis as the symbolic definition of a resource can occur after its use.

Previous merge with main triggered asan memory leaks as shown in https://github.com/llvm/llvm-project/pull/82022#issuecomment-1982288737 by @fmayer. The leak was caused by SmallVector's grow functionality that got invoked as the variadic argument count exceeds the local storage limit (and results in heap allocation). However, MCTargetExpr and all its derived classes do not deallocate meaning any normal heap allocation within AMDGPUVariadicMCExpr does not have a paired deallocation. As a solution, the storage used by AMDGPUVariadicMCExpr for its variadic MCExpr container is now bump-allocated as well through the MCContext allocation.

Relands #82022 with fixes

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUMCExpr.cpp [hunk 1]
```diff
@@ -20,7 +20,13 @@ using namespace llvm;
 const AMDGPUVariadicMCExpr *
 AMDGPUVariadicMCExpr::create(VariadicKind Kind, ArrayRef<const MCExpr *> Args,
                              MCContext &Ctx) {
-  return new (Ctx) AMDGPUVariadicMCExpr(Kind, Args);
+  // Storage for the argument's 'const MCExpr*' allocated through MCContext new placement which means that AMDGPUVariadicMCExpr objects and all of its contents will now be allocated through MCContext new placement.
+  //
+  // Will result in an asan failure if allocated on the heap (e.g., through SmallVector's grow).
+  const MCExpr **CtxArgs = new (Ctx) const MCExpr*[Args.size()];
+  for (size_t i = 0; i < Args.size(); ++i)
+    CtxArgs[i] = Args[i];
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUMCExpr.cpp]: Any special low level management of this should be down in the constructor; the SmallVector's allocator should be the Context's allocator

Comment (JanekvO) [reply llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUMCExpr.cpp]: I was hoping so as well, but when I looked into it `SmallVector` did not have any way to provide an alternate allocator nor does `MCContext` provide the llvm Allocator interface. The `BumpPtrAllocator` used in `MCContext` is also not directly accessible leaving not that many options.

The `AMDGPUVariadicMCExpr::create` function is currently the only place that does the direct creation of the AMDGPUVariadicMCExpr and uses `MCContext` placement new so it seems the most fitting place for this copy code. For now I've replaced the manual copy with `std::copy`.

Do let me know if I'm missing something obvious in terms of allocator behaviour in `SmallVector` or `ArrayRef`.

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUMCExpr.cpp]: Can you just use a regular old array, plus Ctx.allocate?

Comment (JanekvO) [reply llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUMCExpr.cpp]: > regular old array

Sorry, I'm not sure I understand for what aspect I'd use the array. I can, however, replace the placement new approach with one that uses Ctx.allocate.

File: llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUMCExpr.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                             MCContext &Ctx) {
+  // Storage for the argument's 'const MCExpr*' allocated through MCContext new
+  // placement which means that AMDGPUVariadicMCExpr objects and all of its
+  // contents will now be allocated through MCContext new placement.
+  //
+  // Will result in an asan failure if allocated on the heap (e.g., through
+  // SmallVector's grow).
+  const MCExpr **CtxArgs = static_cast<const MCExpr **>(
+      Ctx.allocate(sizeof(const MCExpr *) * Args.size()));
+  std::uninitialized_copy(Args.begin(), Args.end(), CtxArgs);
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUMCExpr.cpp:31]: This should be done in the AMDGPUVariadicMCExpr constructor. Also, I see MCContext does have operator new/delete wrappers around allocate.

Plus you're missing the deallocate which needs to go in the destructor

Comment (JanekvO) [reply llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUMCExpr.cpp:31]: I'm following convention for MCExpr to the extend that I can for AMDGPUVariadicMCExpr: no (target specific) MCExpr is provided the MCContext object and all explicitly have `create` functions to ensure the MCExpr object is allocated through MCContext (i.e., through the MCContext placement new). Additionally, [MCTargetExpr requires subclasses to have trivial destructors as it is never deallocated](https://github.com/llvm/llvm-project/blob/ffe41819e58365dfbe85a22556c0d9d284e746b9/llvm/include/llvm/MC/MCExpr.h#L654-L655). This is further reinforced as MCContext's deallocate function has an [empty body](https://github.com/llvm/llvm-project/blob/ffe41819e58365dfbe85a22556c0d9d284e746b9/llvm/include/llvm/MC/MCContext.h#L851).

Of course, I could be missing something but I do believe this is as good as it gets for supporting variadic arguments in a (target specific) MCExpr.

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUMCExpr.cpp:31]: The regular non-target MCExpr's all have an MCContext argument, I don't see why target MCExprs shouldn't have one. It's not clear to me exactly what that comment means by "trivial". I would presume a destructor that calls a deallocate function which does nothing to be trivial. I think it's much better to properly express the alloc/dealloc in the constructor/destructor than relying on whatever the allocation behavior is today

Comment (JanekvO) [reply llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUMCExpr.cpp]: > The regular non-target MCExpr's all have an MCContext argument
I believe it's only their `create` functions that have it but it is not forwarded beyond that (i.e., to the MCExpr constructor).

> It's not clear to me exactly what that comment means by "trivial"
I assume it's referring to the C++ standard definition of trivial destructors but I don't believe the MCTargetExpr class itself has a trivial destructor according to those rules.

I'm okay with going off-script with the AMDGPUVariadicMCExpr and adding the MCContext as member + user defined destructor, it doesn't trigger the asan leak either way.

**Review by arsenm — APPROVED** [review]

**Pull request merged**